File: packages/types/src/llm/response.ts
/**
 * LLM Response Type Definition
 */

import type { ID, Timestamp, Metadata } from "../common.js";
import type { LLMMessage, LLMToolCall } from "../message/index.js";
import type { LLMUsage } from "./usage.js";

/**
 * LLM response result types (integrating choices and finishReason)
 */
export interface LLMResult {
  /** Response ID */
  id: ID;
  /** Model name */
  model: string;
  /** Response Content Text */
  content: string;
  /** The complete LLMMessage object */
  message: LLMMessage;
  /** Tool Call Array */
  toolCalls?: LLMToolCall[];
  /** Token usage */
  usage?: LLMUsage;
  /** Reason for completion */
  finishReason: string;
  /** Response time (milliseconds) */
  duration: Timestamp;
  /** response metadata */
  metadata?: Metadata;
  /**
   * Thinking/reasoning content
   *
   * Used to store the thought process or reasoning content of a model
   * - OpenAI: reasoning_content (DeepSeek R1, o1 and other reasoning models)
   * - Anthropic: thinking (Claude extended thinking)
   * - Gemini: thoughts (Gemini thinking models)
   */
  reasoningContent?: string;
  /**
   * Thinking/reasoning about the number of Token
   *
   * Used to differentiate between thinking token and regular token billing
   */
  reasoningTokens?: number;
  /**
   * Streaming statistical information
   *
   * Records performance data during streaming generation
   */
  streamStats?: {
    /** Total number of received streaming blocks */
    chunkCount: number;
    /** Time from request to first packet arrival (milliseconds) */
    timeToFirstChunk: number;
    /** Time from first packet to last packet (milliseconds) */
    streamDuration: number;
    /** Total time (in milliseconds) from request sent to end of stream */
    totalDuration: number;
  };
}

File: crates/foundation/wf-types/src/llm/response.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct LlmWarning {
    pub code: String,
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<super::super::Metadata>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct StreamStats {
    pub chunk_count: u32,
    pub time_to_first_chunk: i64,
    pub stream_duration: i64,
    pub total_duration: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct LlmResult {
    pub id: Option<String>,
    pub model: String,
    pub content: Option<String>,
    pub message: super::super::message::Message,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_calls: Option<Vec<super::super::message::LlmToolCall>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub usage: Option<super::TokenUsageStats>,
    pub finish_reason: Option<String>,
    pub duration: i64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reasoning_content: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reasoning_tokens: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<super::super::Metadata>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stream_stats: Option<StreamStats>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub warnings: Option<Vec<LlmWarning>>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ChatResponse {
    pub id: Option<String>,
    pub model: String,
    pub choices: Vec<ChatChoice>,
    pub usage: Option<super::TokenUsageStats>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ChatChoice {
    pub index: u32,
    pub message: serde_json::Value,
    pub finish_reason: Option<String>,
}
